### PR TITLE
[wptserve] Rework header encoding/decoding in Request

### LINKED ISSUE #13204: Many tests in /cookies/http-state consistently time out on Taskcluster but not on BuildBot
https://staging.wpt.fyi/results/cookies/http-state?sha=b040cb2a07&product=firefox[experimental,taskcluster]&product=firefox[buildbot,experimental]&diff=true
https://staging.wpt.fyi/results/cookies/http-state?sha=b040cb2a07&product=chrome[experimental,taskcluster]&product=chrome[buildbot,experimental]&diff=true

Presumably, the tests are a bit slow. I'll try to give them long timeout.

### PR DESCRIPTION
**Pull request by Hexcles**
This PR relands the reverted #13248 from @Ms2ger , but with some tweaks on how to encode/decode the headers. See individual commits for details.

Fixes #13204 .

### CODE REVIEW

File: tools/wptserve/tests/functional/test_request.py
Comment (gsnedders) [inline tools/wptserve/tests/functional/test_request.py]: I think we should test more than just UTF-8 (given this would still pass with `header.decode('utf-8')`).

File: tools/wptserve/wptserve/ranges.py
Comment (gsnedders) [inline tools/wptserve/wptserve/ranges.py]: Let's put this in a try block, with `except UnicodeDecodeError: raise HTTPException(400, "Non-ASCII range header value")`, rather than returning a somewhat mysterious error.

File: tools/wptserve/wptserve/request.py
Comment (gsnedders) [inline tools/wptserve/wptserve/request.py]: Can you write a docstring to actually make it clear what this is and what this is for?

File: tools/wptserve/wptserve/request.py
Comment (gsnedders) [inline tools/wptserve/wptserve/request.py]: Why the behaviour change here?

File: tools/wptserve/wptserve/request.py
Comment (gsnedders) [inline tools/wptserve/wptserve/request.py]: Given the historic differences here, I think we should give the username/password as bytes? Certainly we shouldn't do something irreversible (and throw in the invalid UTF-8 case).

File: tools/wptserve/wptserve/request.py
Comment (Hexcles) [reply tools/wptserve/wptserve/request.py]: Because `cgi.FieldStorage` needs to parse the headers and expect the values to be string not bytes. Since `self.readers` is read-only, I figure it would be safe to pass in the original `raw_headers` here.

**Comment by Hexcles** [comment]
A few clarifications from the discussions on IRC regarding whether we should encode the header values back to bytes (i.e. `headers = [maybeencode(items[header])]`): https://w3.logbot.info/testing/20180928#c444317

First of all, regardless of the approach (whether to encode and which encoding to use), this WILL NOT affect existing tests (namely custom handlers) at all. Currently, WPT only really works with Python 2, and in Python 2 headers (both names and values) have `str` type, which is a "binary" type, so `maybeencode` is a no-op.

People reasonably want `wptserve` itself to work in Python 3, and this is where the change has effects. I would hope Python 3 still gives us headers in binary type (i.e. `bytes`), as Python 2, but Python 3 decides to decode the headers in iso-8859-1 perhaps for better compatibility. I think the sane choice here is to return `bytes`, which requires re-encoding, as HTTP RFC doesn't really say what encoding headers *must* use and encoding non-ASCII characters in iso-8859-1 is broken. The downside is that potentially more places that manipulate headers need to be changed (e.g. literal operands of string operations need to be `bytes`).

Now, since many custom handlers don't work in Python 3 in the first place, I'm inclined to to `maybeencode` the header values (i.e. return `bytes` in Python 3).

**Comment by Hexcles** [comment]
@gsnedders This PR is ready for review again. I addressed all your comments but one (the `cgi.FieldStorage` one; please see my reply there).

**Comment by annevk** [comment]
> and encoding non-ASCII characters in iso-8859-1 is broken

It's not? Using bytes is perhaps purer in a way, but if you make them strings that is the way to go. Anything else would lead to information loss.

**Comment by Ms2ger** [comment]
I suspect that supporting only `binary_type` in this API would be the most straightforward approach.

**Comment by Hexcles** [comment]
@Ms2ger you mean keys should also be binary type? Yeah I can do that. It might be break a few more things (in Python 3 only) but it's probably more consistent.

**Comment by gsnedders** [comment]
@Ms2ger @Hexcles I have no strong preference; in principle header names are strictly ASCII only, though I don't think impls enforce that.

**Comment by Hexcles** [comment]
@Ms2ger @gsnedders there's one complication here. Technically, a byte string doesn't have the `lower` method. Yet we usually wish header names to be case-insensitive and we implement it by always converting the keys to lowercase both during storage and lookup.

**Comment by gsnedders** [comment]
@Hexcles Yes it does? `bytes.lower` does ASCII lowercasing on both Py2/3.

**Comment by Hexcles** [comment]
@gsnedders duh my bad. Yeah let me update the PR.

File: tools/wptserve/tests/functional/base.py
Comment (gsnedders) [inline tools/wptserve/tests/functional/base.py]: I wonder if we want to not encode there, so we can test with non-UTF-8 values?

File: tools/wptserve/tests/functional/test_pipes.py
Comment (gsnedders) [inline tools/wptserve/tests/functional/test_pipes.py]: I presume this is residual from debugging? :)

File: tools/wptserve/tests/functional/test_request.py
Comment (gsnedders) [reply tools/wptserve/tests/functional/test_request.py]: I'd rather this be done with `pytest.mark.parametrize` so they run as separate tests, FWIW.

File: tools/wptserve/wptserve/pipes.py
Comment (gsnedders) [inline tools/wptserve/wptserve/pipes.py]: ~~Can we just avoid the roundtrip here?~~ No, we can't, can you add a comment saying *why* we can't? (i.e., the type of escape_func)

File: tools/wptserve/wptserve/request.py
Comment (gsnedders) [reply tools/wptserve/wptserve/request.py]: `self.raw_headers` is the headers as `bytes`, no? What type does cgi.FieldStorage expect? `str` on both Py2/3, no?

File: tools/wptserve/wptserve/request.py
Comment (gsnedders) [inline tools/wptserve/wptserve/request.py]: "Besides" seem a bit clumsy here. Probably just make this "and"?

File: tools/wptserve/tests/functional/test_pipes.py
Comment (gsnedders) [inline tools/wptserve/tests/functional/test_pipes.py]: Is there any reason not to use a binary string here? format still works, no?

File: tools/wptserve/wptserve/pipes.py
Comment (gsnedders) [inline tools/wptserve/wptserve/pipes.py]: This is a separate fix, yes? Do we have any tests for this? ^_^

Also: don't we want both of the header args to be binary strings? I can't see what goes on here looking quickly? But I guess the Response side probably makes sense to do so separately.

File: tools/wptserve/tests/functional/test_pipes.py
Comment (Hexcles) [reply tools/wptserve/tests/functional/test_pipes.py]: Oops. Removed.

File: tools/wptserve/wptserve/pipes.py
Comment (Hexcles) [reply tools/wptserve/wptserve/pipes.py]: I didn't modify this file; the change was cherry-picked from #11769 by @Ms2ger .

File: tools/wptserve/tests/functional/test_pipes.py
Comment (Hexcles) [reply tools/wptserve/tests/functional/test_pipes.py]: No, `bytes` doesn't have `format` method. We could use `%`, but it'd be a bit clumsy as we need to repeat the port number a few times.

File: tools/wptserve/tests/functional/test_request.py
Comment (Hexcles) [reply tools/wptserve/tests/functional/test_request.py]: This is a neat feature! However, it looks like `pytest.mark.parametrize` [does not work](https://docs.pytest.org/en/latest/unittest.html#pytest-features-in-unittest-testcase-subclasses) with `unittest.TestCase`.

File: tools/wptserve/tests/functional/base.py
Comment (Hexcles) [reply tools/wptserve/tests/functional/base.py]: Done. Also added a test case (in Shift JIS).

**Comment by andypaicu** [comment]
Just as an fyi guys I'm also waiting on this re-land as I suspect the revert broke the "wpt / content-security-policy / generic / only-valid-whitespaces-are-allowed.html" test.

There hasn't been any movement on this for a week and I just want to make sure that this will eventually get merged.

https://wpt.fyi/results/content-security-policy/generic/only-valid-whitespaces-are-allowed.html?label=experimental

File: tools/wptserve/tests/functional/test_request.py
Comment (gsnedders) [reply tools/wptserve/tests/functional/test_request.py]: Oh, I missed your response here!

Ergh, another reason to move away from `unittest` entirely, but that probably shouldn't block this PR.

**Review by gsnedders — APPROVED** [review]

**Comment by gsnedders** [comment]
@Hexcles @Ms2ger do one of you want to rewrite history to something saner?

**Comment by Ms2ger** [comment]
If @Hexcles hasn't done it by tomorrow, I'll squash them all together.

**Comment by Hexcles** [comment]
Thanks for the review, @gsnedders . (The comment threads do look quite confusing...)

I just cleaned up the history and will merge the rebase-merge PR once Travis passes.

**Pull request merged**
